csrc/runtime/allocations.cpp

```diff
@@ -336,6 +336,109 @@ std::vector<GlobalBufferInfo> getBufferInfos(
 
 namespace {
 
+// Helper function to see if the dimension of a tensor
+// which are being merged via the vector new_shape are
+// contiguous or not. This function only allows for two dims
+// to be merged - and the -1 in new_shape indicates the
+// dimension to be merged with the next one.
+bool areDimsToBeMergedContiguous(
+    const at::Tensor& tensor,
+    const std::vector<int64_t>& new_shape) {
+  // Assert that new_shape contains at most one -1
+  NVF_ERROR(
+      std::count(new_shape.begin(), new_shape.end(), -1) <= 1,
+      "new_shape can contain at most one -1");
+
+  auto it = std::find(new_shape.begin(), new_shape.end(), -1);
+  if (it == new_shape.end()) {
+    return true;
+  }
+  size_t merge_dim = std::distance(new_shape.begin(), it);
+
+  if (merge_dim == new_shape.size() - 1) {
+    return true;
+  }
+
+  auto strides = tensor.strides();
+  auto sizes = tensor.sizes();
+
+  // Get the size and stride at merge_dim and merge_dim + 1
+  NVF_ERROR(merge_dim < sizes.size(), "merge_dim out of bounds for sizes");
+  NVF_ERROR(
+      merge_dim + 1 < sizes.size(), "merge_dim+1 out of bounds for sizes");
+  auto size_outer = sizes[merge_dim];
+  auto stride_outer = strides[merge_dim];
+  auto size_inner = sizes[merge_dim + 1];
+  auto stride_inner = strides[merge_dim + 1];
+
+  // We are just squeezing out a dim.
+  // For example we can call view on the tensor with shape
+  // [1, 1536, 2, 128] and strides [393216, 1, 196608, 1536]
+  // Then we should be able to merge the dims [1, 1536].
+  if (size_outer == 1 || size_inner == 1) {
+    return true;
+  }
+
+  if (stride_outer == size_inner * stride_inner) {
+    return true;
+  }
+
+  return false;
+}
+
+// Helper function to compute the shape and strides needed for merging
+// dimensions in a tensor. Used when merging dimensions in the allocation domain
+// that may have been permuted, making them non-contiguous.
+//
+// For example, if a tensor has shape [8, 16, 4, 32, 4] and strides [8192, 512,
+// 4, 16, 1] after a permute, and we want to merge the [4, 32] dimensions,
+// view() would fail since they are non-contiguous. This function computes the
+// correct shape and strides to use with as_strided() instead.
+//
+// Args:
+//   tensor: The input tensor whose dimensions will be merged
+//   new_shape: Target shape with -1 indicating dimensions to merge with the
+//   next dim
+//
+// Returns:
+//   A pair containing:
+//   - The final shape after merging dimensions
+//   - The strides needed for as_strided() to properly merge the dimensions
+std::pair<std::vector<int64_t>, std::vector<int64_t>>
+getShapeAndStrideAfterDimMerged(
+    const at::Tensor& tensor,
+    const std::vector<int64_t>& new_shape) {
+  // Copy tensor shape into std::vector<int64_t>
+  std::vector<int64_t> tensor_shape_vec(
+      tensor.sizes().begin(), tensor.sizes().end());
+  std::vector<int64_t> tensor_new_shape;
+  size_t i = 0;
+  while (i < new_shape.size()) {
```

Comment:
nitpick: we are not using `i` out side of the loop, this looks like a simple for loop thing to me.